File: src/main/kotlin/uniquePaths/UniquePaths.kt
package uniquePaths

class UniquePaths {
    fun uniquePaths(m: Int, n: Int): Int {
        val min = min(m - 1, n - 1)
        val sum = m + n - 2
        var result = 1L
        for (i in 0 until min) {
            result *= (sum - i)
            result /= (i + 1)
        }
        return result.toInt()
    }

    private fun min(m: Int, n: Int): Int {
        return if (m < n) m else n
    }
}

File: src/main/python/uniquePaths/UniquePaths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        def factorial(memo, m):
            if m not in memo:
                memo[m] = factorial(memo, m - 1) * m
            return memo[m]

        memo = {0: 1, 1: 1}
        target_m, target_n = m - 1, n - 1
        factorial_m_n = factorial(memo, target_m + target_n)
        return factorial_m_n // memo[target_m] // memo[target_n]

    def test(self):
        tests = [
            [3, 7, 28],
            [3, 2, 3],
        ]
        for test in tests:
            m, n, expect_result = test
            result = self.uniquePaths(m, n)
            print(result)
            print(result == expect_result)


if __name__ == '__main__':
    Solution().test()
